### PR TITLE
More Jinja IP filters

### PR DESCRIPTION
**Pull request by mirceaulinic**
### What does this PR do?

Defines few more network related util functions and links them + existing ones for Jinja filters.

Allows doing operations such:

```jinja
{% if my_string | is_ipv4 -%}
```

```jinja
{% if my_string | is_ipv4('multicast') -%}
```

```jinja
{{ my_list | ipv4 }}
``` 
to display IPv4-only addresses from ```my_list``` etc.

### CODE REVIEW

File: salt/utils/network.py [hunk 1]
```diff
@@ -199,31 +200,239 @@ def ip_to_host(ip):
...
-def is_ipv4(ip):
-    '''
-    Returns a bool telling if the value passed to it was a valid IPv4 address
-    '''
+def _ip_options_global(ip_obj, version):
+    return not ip_obj.is_private
+
+
+def _ip_options_multicast(ip_obj, version):
+    return ip_obj.is_loopback
```
Comment (silenius) [inline salt/utils/network.py]: I thin you want `return ip_obj.is_multicast` here ..?

File: salt/utils/network.py [hunk 2]
```diff
@@ -199,31 +200,239 @@ def ip_to_host(ip):
...
+        return None
+
     try:
-        return ipaddress.ip_address(ip).version == 4
+        ip_obj = ipaddress.ip_address(ip)
     except ValueError:
-        return False
+        # maybe it is an IP network
+        try:
+            ip_obj = ipaddress.ip_interface(ip)
```
Comment (silenius) [inline salt/utils/network.py]: be careful of any side effects here if this function is used elsewhere ... (ip address vs ip network)

File: salt/utils/network.py [hunk 3]
```diff
@@ -199,31 +200,239 @@ def ip_to_host(ip):
...
+        except ValueError:
+            # nope, still not :(
+            return None
```
Comment (silenius) [inline salt/utils/network.py]: `False` was returned previously, this may break things

File: salt/utils/network.py [hunk 4]
```diff
@@ -199,31 +200,239 @@ def ip_to_host(ip):
...
+            fun(addr, options=options)
+            for addr in value
+            if fun(addr, options=options) is not None
+        ]
+    return None
+
+
+def ipv4(value, options=None):
+    '''
+    Filters a list and returns IPv4 values only.
```
Comment (silenius) [inline salt/utils/network.py]: you really need another *ipv4* function for that ..? can't you just use a list comprehension ?

File: salt/utils/network.py [hunk 5]
```diff
@@ -199,31 +200,239 @@ def ip_to_host(ip):
...
+    '''
+    return _ipv_filter(value, 4, options=options)
+
+
+def ipv6(value, options=None):
+    '''
+    Filters a list and returns IPv6 values only.
```
Comment (silenius) [inline salt/utils/network.py]: same here ...

File: salt/utils/network.py [hunk 6]
```diff
@@ -199,31 +200,239 @@ def ip_to_host(ip):
...
+def _ip_options_global(ip_obj, version):
+    return not ip_obj.is_private
+
+
+def _ip_options_multicast(ip_obj, version):
+    return ip_obj.is_loopback
+
+
+def _ip_options_loopback(ip_obj, version):
+    return ip_obj.is_multicast
```
Comment (mirceaulinic) [inline salt/utils/network.py]: Dammit, will fix that!

File: salt/utils/network.py (hunk 2 above)
Comment (mirceaulinic) [inline salt/utils/network.py]: That's correct: previously a call such ```172.17.17.1/32``` would return False, although is the equivalent for ```172.17.17.1``` for which it returns True. But you are right, we should avoid that - will correct it tomorrow!

File: salt/utils/network.py (hunk 3 above)
Comment (mirceaulinic) [inline salt/utils/network.py]: Will fix, as stated above.

Comment (mirceaulinic) [inline salt/utils/network.py]: I opted to define separate functions ending in ```_filter``` which are used for jinja filters, whilst the old functions remain the same without changing the behaviour.

**Comment by mirceaulinic** [comment]
This one is ready IMO

**Comment by cachedout** [comment]
Thanks so much, @mirceaulinic. Awesome work as always!

**Pull request merged**